### PR TITLE
New resources for Log analytics datasources

### LINKED ISSUE #3182: Configure Data Sources in a Log Analytics Workspace
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I'd like to be able to configure Data Sources on a Log Analytics Workspace

### New or Affected Resource(s)

* azurerm_log_analytics_workspace

and/or:

* azurerm_log_analytics_workspace_iis_logs
* azurerm_log_analytics_workspace_linux_performance_collection
* azurerm_log_analytics_workspace_linux_performance_object
* azurerm_log_analytics_workspace_linux_syslog
* azurerm_log_analytics_workspace_linux_syslogCollection
* azurerm_log_analytics_workspace_windows_event
* azurerm_log_analytics_workspace_windows_performance_counter

Note: There may be other data sources and properties that can be configured.

### Potential Terraform Configuration

Inline example of configuring "datasource" properties in a Log Analytics workspace

```hcl
resource "azurerm_log_analytics_workspace" "ws" {
  name                = "Demo"
  location            = "US East"
  resource_group_name = "Some-Resource-Group"
  sku                 = "Standard"

  windows_performance_counter { 
    object_name = "LogicalDisk"
    instance_name = "*"
    interval_seconds = 10
    counter_name = "Disk Writes/sec"
  }

  windows_performance_counter { 
    object_name = "LogicalDisk"
    instance_name = "*"
    interval_seconds = 10
    counter_name = "Free Megabytes"
  }

  linux_performance_collection {
    state = "Enabled"
  }

  linux_performance_object {
    performance_counters = ["% Processor Time", "% Privileged Time"]
    object_name = "Processor"
    instance_name = "*"
    interval_seconds = 10
  }
}
```

Configure "datasource" using a separate resource for a Log Analytics workspace.

```hcl
resource "azurerm_log_analytics_workspace_iis_logs" "iis_logs" {
  name                = "iis_logs"
  resource_group_name = "Some-Resource-Group"

  state = "OnPremiseEnabled"
}

resource "azurerm_log_analytics_workspace_linux_performance_collection" "lin_perf_collection" {
  name                = "lin_perf_collection"
  resource_group_name = "Some-Resource-Group"

  state = "Enabled"
}

resource "azurerm_log_analytics_workspace_linux_performance_object" "lin_processor_performance" {
  name                = "lin_processor_performance"
  resource_group_name = "Some-Resource-Group"

  performance_counters = ["% Processor Time", "% Privileged Time"]
  object_name = "Processor"
  instance_name = "*"
  interval_seconds = 10
}

resource "azurerm_log_analytics_workspace_linux_syslog_collection" "lin_syslog_collection" {
  name                = "lin_syslog_collection"
  resource_group_name = "Some-Resource-Group"

  state = "Enabled"
}

resource "azurerm_log_analytics_workspace_linux_syslog_collection" "lin_syslog_collection" {
  name                = "lin_syslog_collection"
  resource_group_name = "Some-Resource-Group"

  syslog_name = "kern"
  severities = ["emerg", "alert", "crit", "err", "warning"]
}

resource "azurerm_log_analytics_workspace_windows_event" "win_event" {
  name                = "win_event"
  resource_group_name = "Some-Resource-Group"
  
  eventLogName = "Application"
  eventTypes = ["Error", "Warning"]
}

resource "azurerm_log_analytics_workspace_windows_performance_counter" "win_disk_writes" {
  name                = "win_disk_writes"
  resource_group_name = "Some-Resource-Group"

  object_name = "LogicalDisk"
  instance_name = "*"
  interval_seconds = 10
  counter_name = "Disk Writes/sec"
}
```

### PR DESCRIPTION
**Pull request by PriyankaRanganath**
Add support for Log analytics workspace data source resources:

1) `azurerm_log_analytics_datasource_iis_logs`
2) `azurerm_log_analytics_datasource_linux_performance_collection`
3) `azurerm_log_analytics_datasource_linux_performance_object`
4) `azurerm_log_analytics_datasource_linux_syslog_collection`
5) `azurerm_log_analytics_datasource_linux_syslog`

fixes #3182

### CODE REVIEW

**Comment by PriyankaRanganath** [comment]
Hello, 
can anyone review my PR ?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go:60]: Could we change this attribute to be `on_premise_enabled` and the value be a bool? WDYT?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go:115]: Super minor...
```suggestion
		return fmt.Errorf("unable to read ID for Log Analytics DataSource IIS logs %q (Resource Group %q)", name, resourceGroup)
```

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go:60]: Same here as above, can we give this a more meaningful name here, something like `datasource_enabled` or even just `enabled` would work and have the value be a bool?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_collection_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_collection_resource.go:60]: Same here, can we change this to something meaningful or just `enabled`?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:75]: Can we get these values from the SDK rather than hand typing the values?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:87]: Same here, can we pull these values from the SDK?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:89]: You should be able to remove this if you are pulling the values from the SDK?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:212]: can we add a nil or empty value check here?

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (WodansSon) [inline azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:220]: Same here, can we add a nil or empty value check here?

File: website/docs/r/log_analytics_datasource_iis_logs.html.markdown
Comment (WodansSon) [inline website/docs/r/log_analytics_datasource_iis_logs.html.markdown:36]: Can we update this to reflect the new value per the comment above? 
```suggestion
  on_premise               = true
```

File: website/docs/r/log_analytics_datasource_iis_logs.html.markdown
Comment (WodansSon) [inline website/docs/r/log_analytics_datasource_iis_logs.html.markdown:50]:
```suggestion
* `on_premise` - (Required) Specifies the start/stop state for collection of performance data from connected IIS servers. Possible values include `true` and `false`. 
```

File: website/docs/r/log_analytics_datasource_linux_performance_collection.html.markdown
Comment (WodansSon) [inline website/docs/r/log_analytics_datasource_linux_performance_collection.html.markdown:36]:
```suggestion
  enabled             = true
```

File: website/docs/r/log_analytics_datasource_linux_performance_collection.html.markdown
Comment (WodansSon) [inline website/docs/r/log_analytics_datasource_linux_performance_collection.html.markdown:50]:
```suggestion
* `enabled` - (Required) Specifies the start/stop state for collection of performance data from connected linux computers. Possible values include `true` and `false`.
```

File: website/docs/r/log_analytics_datasource_linux_syslog_collection.html.markdown
Comment (WodansSon) [inline website/docs/r/log_analytics_datasource_linux_syslog_collection.html.markdown:36]:
```suggestion
  enabled             = true
```

**Review by WodansSon — CHANGES_REQUESTED** [review]
@PriyankaRanganath, thank you so much for opening this PR, this mostly LGTM, however I left a few comments to address. Once those are taken care of we can get this merged! 🚀

File: azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go]: Agreed @WodansSon, I have made changes likewise

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:212]: Added nil check

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:223]: Added

File: azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go]: Resolved

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_collection_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_collection_resource.go]: Added "enabled" instead of "state"

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go]: Added "enabled" instead of "state"

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:75]: There is no enum support for data source properties in SDK. I referenced the powershell documentation from here: https://docs.microsoft.com/en-us/powershell/module/az.operationalinsights/new-azoperationalinsightslinuxsyslogdatasource?view=azps-4.7.0#examples

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:87]: Same as above. No documentation for severities. I referenced ARM documentation from here: https://docs.microsoft.com/en-us/azure/azure-monitor/samples/resource-manager-workspace#collect-syslog

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_syslog_resource.go:89]: Due to no enum support from az-sdk-for-go, can not remove them

File: website/docs/r/log_analytics_datasource_iis_logs.html.markdown
Comment (PriyankaRanganath) [reply website/docs/r/log_analytics_datasource_iis_logs.html.markdown]: Resolved to reflect new schema properties

File: website/docs/r/log_analytics_datasource_iis_logs.html.markdown
Comment (PriyankaRanganath) [reply website/docs/r/log_analytics_datasource_iis_logs.html.markdown]: Resolved to reflect new schema properties

File: website/docs/r/log_analytics_datasource_linux_syslog_collection.html.markdown
Comment (PriyankaRanganath) [reply website/docs/r/log_analytics_datasource_linux_syslog_collection.html.markdown]: Resolved to reflect new schema properties

**Comment by PriyankaRanganath** [comment]
Hi @WodansSon ,

Resolved comments as requested

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @PriyankaRanganath, while overall this looks great it appears the SDK/swagger is incomplete and needs to be updated and unfortunately we need the SDK to define these types else they can change on Azure's side without correctly handling semantic versioning - as such these types need to be present in the Azure SDK/Swagger. There is an open issue [here](https://github.com/Azure/azure-rest-api-specs/issues/9072) and until this is fixed i'm not sure we can merge this

**Comment by katbyte** [comment]
linked issue: https://github.com/Azure/azure-rest-api-specs/issues/9072

File: azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_iis_logs_resource.go]: Done

File: azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go
Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go:87]: Hi @katbyte , there is no support for SDK or even official documentation from Microsoft yet. Is it ok to still merge this PR ?
I am trying to get hold of Microsoft prod team but as mentioned there is no ETA on it yet

Comment (katbyte) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go:87]: Please see the linked issue below for more details, but we typically refrain from merging anything that is not currently supported by the SDK as it can become problematic down the line.

Comment (PriyankaRanganath) [reply azurerm/internal/services/loganalytics/log_analytics_datasource_linux_performance_collection_resource.go:87]: Thanks @katbyte ! will wait for SDK and official documentation release

**Comment by katbyte** [comment]
@PriyankaRanganath - given this is blocked on the swagger issue, i am going to close it. Once the service team has addressed the missing swagger & SDK supports these properties we can reopen this PR and move it forward.

**Pull request closed without merging**
